File: blue-score-layers-patterns-core/src/blue/score/layers/patterns/core/PatternsLayerGroup.java
/*
 * blue - object composition environment for csound
 * Copyright (C) 2012
 * Steven Yi <stevenyi@gmail.com>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */
package blue.score.layers.patterns.core;

import blue.CompileData;
import blue.noteProcessor.NoteProcessorChain;
import blue.noteProcessor.NoteProcessorException;
import blue.score.ScoreGenerationException;
import blue.score.layers.Layer;
import blue.score.layers.LayerGroup;
import blue.score.layers.LayerGroupDataEvent;
import blue.score.layers.LayerGroupListener;
import blue.soundObject.*;
import blue.utility.ScoreUtilities;
import electric.xml.Element;
import electric.xml.Elements;
import java.util.ArrayList;
import java.util.Map;
import java.util.Vector;

/**
 *
 * @author stevenyi
 */
public class PatternsLayerGroup extends ArrayList<PatternLayer>
        implements LayerGroup<PatternLayer> {

    private transient Vector<LayerGroupListener> layerGroupListeners = null;
    private int patternBeatsLength = 4;
    private String name = "Patterns Layer Group";
    private NoteProcessorChain npc;

    public PatternsLayerGroup() {
        npc = new NoteProcessorChain();
    }

    public PatternsLayerGroup(PatternsLayerGroup plg) {
        super(plg.size());
        patternBeatsLength = plg.patternBeatsLength;
        name = plg.name;
        npc = new NoteProcessorChain(plg.npc);

        for (PatternLayer pl : plg) {
            add(pl.deepCopy());
        }
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public void setName(String name) {
        this.name = name;
    }

    @Override
    public NoteProcessorChain getNoteProcessorChain() {
        return npc;
    }

    public void setNoteProcessorChain(NoteProcessorChain npc) {
        this.npc = npc;
    }

    public int getPatternBeatsLength() {
        return patternBeatsLength;
    }

    public void setPatternBeatsLength(int patternBeatsLength) {
        this.patternBeatsLength = patternBeatsLength;
    }

    @Override
    public boolean hasSoloLayers() {
        for (PatternLayer layer : this) {
            if (layer.isSolo()) {
                return true;
            }
        }
        return false;
    }

    @Override
    public NoteList generateForCSD(CompileData compileData, double startTime, double endTime, boolean processWithSolo) throws ScoreGenerationException {

        NoteList noteList = new NoteList();

        if (processWithSolo) {
            for (PatternLayer patternLayer : this) {
                if (patternLayer.isSolo()) {
                    if (!patternLayer.isMuted()) {
                        noteList.merge(patternLayer.generateForCSD(compileData,
                                startTime, endTime,
                                patternBeatsLength));
                    }
                }
            }
        } else {
            for (PatternLayer patternLayer : this) {
                if (!patternLayer.isMuted()) {
                    noteList.merge(patternLayer.generateForCSD(compileData,
                            startTime, endTime,
                            patternBeatsLength));
                }
            }
        }
        try {
            noteList = processNotes(compileData, noteList, startTime, endTime);
        } catch (NoteProcessorException ex) {
            throw new ScoreGenerationException(ex);
        }

        return noteList;
    }

    private NoteList processNotes(CompileData compileData, NoteList nl, double start, double endTime) throws NoteProcessorException {

        NoteList retVal = null;

        ScoreUtilities.applyNoteProcessorChain(nl, this.npc);

        if (start == 0.0f) {
            retVal = nl;
        } else {
            ScoreUtilities.setScoreStart(nl, -start);

            NoteList buffer = new NoteList();
            Note tempNote;

            for (int i = 0; i < nl.size(); i++) {
                tempNote = nl.get(i);

                if (tempNote.getStartTime() >= 0) {
                    buffer.add(tempNote);
                }
            }
            retVal = buffer;
        }

        if (endTime > start) {
            // double dur = endTime - start;

            NoteList buffer = new NoteList();
            Note tempNote;

            for (int i = 0; i < retVal.size(); i++) {
                tempNote = retVal.get(i);

                if (tempNote.getStartTime() <= endTime) {
                    buffer.add(tempNote);
                }
            }
            retVal = buffer;
        }

        return retVal;
    }

    public static PatternsLayerGroup loadFromXML(Element data) throws Exception {
        PatternsLayerGroup layerGroup = new PatternsLayerGroup();

        if (data.getAttribute("name") != null) {
            layerGroup.setName(data.getAttributeValue("name"));
        }

        Elements nodes = data.getElements();
        while (nodes.hasMoreElements()) {
            Element node = nodes.next();
            String nodeName = node.getName();

            if ("patternLayers".equals(nodeName)) {

                Elements patternNodes = node.getElements();

                while (patternNodes.hasMoreElements()) {

                    Element patternNode = patternNodes.next();

                    if ("patternLayer".equals(patternNode.getName())) {
                        layerGroup.add(PatternLayer.loadFromXML(
                                patternNode));
                    }
                }
            } else if ("noteProcessorChain".equals(nodeName)) {
                layerGroup.setNoteProcessorChain(NoteProcessorChain.loadFromXML(
                        node));
            }
        }

        return layerGroup;
    }

    @Override
    public Element saveAsXML(Map<Object, String> objRefMap) {
        Element root = new Element("patternsLayerGroup");
        root.setAttribute("name", name);

        Element patternsNode = root.addElement("patternLayers");

        for (PatternLayer layer : this) {
            patternsNode.addElement(layer.saveAsXML());
        }

        root.addElement(npc.saveAsXML());

        return root;
    }

    @Override
    public PatternLayer newLayerAt(int index) {

        PatternLayer patternLayer = new PatternLayer();

        int insertIndex = index;
        if (index < 0 || index >= this.size()) {
            insertIndex = this.size();
            this.add(patternLayer);
        } else {
            this.add(index, patternLayer);
        }

        ArrayList<Layer> layers = new ArrayList<>();
        layers.add(patternLayer);

        LayerGroupDataEvent lde = new LayerGroupDataEvent(this,
                LayerGroupDataEvent.DATA_ADDED, insertIndex, insertIndex, layers);

        fireLayerGroupDataEvent(lde);

        return patternLayer;
    }

    @Override
    public void removeLayers(int startIndex, int endIndex) {

        ArrayList<Layer> layers = new ArrayList<>();

        for (int i = endIndex; i >= startIndex; i--) {
            PatternLayer patternLayer = this.get(i);
            patternLayer.clearListeners();

            this.remove(i);

            layers.add(patternLayer);
        }

        LayerGroupDataEvent lde = new LayerGroupDataEvent(this,
                LayerGroupDataEvent.DATA_REMOVED, startIndex, endIndex, layers);

        fireLayerGroupDataEvent(lde);

    }

    @Override
    public void pushUpLayers(int startIndex, int endIndex) {
        PatternLayer a = this.remove(startIndex - 1);
        this.add(endIndex, a);

        LayerGroupDataEvent lde = new LayerGroupDataEvent(this,
                LayerGroupDataEvent.DATA_CHANGED, startIndex - 1, endIndex);

        fireLayerGroupDataEvent(lde);
    }

    @Override
    public void pushDownLayers(int startIndex, int endIndex) {
        PatternLayer a = this.remove(endIndex + 1);
        this.add(startIndex, a);

        LayerGroupDataEvent lde = new LayerGroupDataEvent(this,
                LayerGroupDataEvent.DATA_CHANGED, -startIndex, -(endIndex + 1));

        fireLayerGroupDataEvent(lde);
    }

    @Override
    public void onLoadComplete() {
        for (PatternLayer layer : this) {
            SoundObject sObj = layer.getSoundObject();

            if (sObj instanceof OnLoadProcessable) {
                OnLoadProcessable olp = (OnLoadProcessable) sObj;
                if (olp.isOnLoadProcessable()) {
                    try {
                        olp.processOnLoad();
                    } catch (SoundObjectException soe) {
                        throw new RuntimeException(new SoundObjectException(sObj,
                                "Error during on load processing:", soe));
                    }
                }
            }
        }
    }

    @Override
    public void addLayerGroupListener(LayerGroupListener l) {
        if (layerGroupListeners == null) {
            layerGroupListeners = new Vector<>();
        }

        layerGroupListeners.add(l);
    }

    @Override
    public void removeLayerGroupListener(LayerGroupListener l) {
        if (layerGroupListeners != null) {
            layerGroupListeners.remove(l);
        }
    }

    private void fireLayerGroupDataEvent(LayerGroupDataEvent lde) {
        if (layerGroupListeners == null) {
            return;
        }

        for (LayerGroupListener listener : layerGroupListeners) {
            listener.layerGroupChanged(lde);
        }
    }

    public int getMaxPattern() {
        int max = 0;
        for (PatternLayer layer : this) {
            if (layer.getPatternData().getMaxSelected() > max) {
                max = layer.getPatternData().getMaxSelected();
            }
        }
        return max;
    }

    @Override
    public PatternsLayerGroup deepCopyLG() {
        return new PatternsLayerGroup(this);
    }
}

File: packages/blue-data/src/score/patterns/patterns-layer-group.ts
/**
 * PatternsLayerGroup — a group of PatternLayers.
 * Mirrors the Java PatternsLayerGroup class.
 *
 * PatternsLayerGroup holds multiple PatternLayers and a patternBeatsLength
 * that determines the duration of each pattern step. During CSD generation,
 * each PatternLayer repeats its SoundObject at active pattern positions.
 */
import { PatternLayer } from './pattern-layer';
import { LayerGroup } from '../../score/layers/layer-group';
import { NoteProcessorChain } from '../../note-processors/note-processor-chain';
import { NoteList } from '../../sound-objects/note-list';
import { TimeContext } from '../../time/time-context';
import { CompileData } from '../../compile-data';
import { ScoreGenerationException } from '../../score/score-generation-exception';
import { Element } from '../../serialization/xml-reader';
import { ObjRefSaveMap, ObjRefLoadMap } from '../../serialization/obj-ref-map';
import {
  normalizeScoreGenerationOptions,
  type ScoreGenerationOptionsOrSolo,
} from '../score-generation-options';

export class PatternsLayerGroup extends Array<PatternLayer> implements LayerGroup<PatternLayer> {
  static get [Symbol.species](): ArrayConstructor {
    return Array;
  }

  private _name = 'Patterns Layer Group';
  private _patternBeatsLength = 4;
  private _npc = new NoteProcessorChain();

  constructor(other?: PatternsLayerGroup) {
    super();
    if (other) {
      this._name = other._name;
      this._patternBeatsLength = other._patternBeatsLength;
      this._npc = new NoteProcessorChain(other._npc);
      for (const pl of other) {
        this.push(pl.deepCopy());
      }
    }
  }

  // ─── LayerGroup ───

  getName(): string {
    return this._name;
  }
  setName(name: string): void {
    this._name = name;
  }

  getPatternBeatsLength(): number {
    return this._patternBeatsLength;
  }
  setPatternBeatsLength(length: number): void {
    this._patternBeatsLength = length;
  }

  getNoteProcessorChain(): NoteProcessorChain {
    return this._npc;
  }
  setNoteProcessorChain(npc: NoteProcessorChain): void {
    this._npc = npc;
  }

  hasSoloLayers(): boolean {
    return this.some((layer) => layer.isSolo());
  }

  generateForCSD(
    context: TimeContext,
    compileData: CompileData,
    startTime: number,
    endTime: number,
    options?: ScoreGenerationOptionsOrSolo,
  ): NoteList {
    const noteList = new NoteList();
    const processWithSolo = normalizeScoreGenerationOptions(options).processWithSolo ?? false;

    if (processWithSolo) {
      for (const patternLayer of this) {
        if (patternLayer.isSolo() && !patternLayer.isMuted()) {
          const nl = patternLayer.generateForCSD(
            context,
            compileData,
            startTime,
            endTime,
            this._patternBeatsLength,
          );
          noteList.merge(nl);
        }
      }
    } else {
      for (const patternLayer of this) {
        if (!patternLayer.isMuted()) {
          const nl = patternLayer.generateForCSD(
            context,
            compileData,
            startTime,
            endTime,
            this._patternBeatsLength,
          );
          noteList.merge(nl);
        }
      }
    }

    // Apply note processor chain
    return this._npc.apply(noteList);
  }

  saveAsXML(objRefMap?: ObjRefSaveMap): Element {
    const root = new Element('patternsLayerGroup');
    root.setAttribute('name', this._name);
    root.addElement('patternBeatsLength').setText(this._patternBeatsLength.toString());

    const patternsNode = root.addElement('patternLayers');
    for (const layer of this) {
      patternsNode.addElement(layer.saveAsXML(objRefMap));
    }

    root.addElement(this._npc.saveAsXML().setName('noteProcessorChain'));

    return root;
  }

  static loadFromXML(data: Element, objRefMap?: ObjRefLoadMap): PatternsLayerGroup {
    const group = new PatternsLayerGroup();

    const name = data.getAttribute('name');
    if (name) group._name = name;

    const nodes = data.getElements();
    while (nodes.hasMoreElements()) {
      const node = nodes.next();
      if (node.getName() === 'patternBeatsLength') {
        group._patternBeatsLength = parseInt(node.getTextString(), 10);
      } else if (node.getName() === 'patternLayers') {
        const patternNodes = node.getElements();
        while (patternNodes.hasMoreElements()) {
          const patternNode = patternNodes.next();
          if (patternNode.getName() === 'patternLayer') {
            group.push(PatternLayer.loadFromXML(patternNode, objRefMap));
          }
        }
      } else if (node.getName() === 'noteProcessorChain') {
        group._npc = NoteProcessorChain.loadFromXML(node);
      }
    }

    return group;
  }

  // ─── LayerGroup operations ───

  newLayerAt(index: number): PatternLayer {
    const patternLayer = new PatternLayer();

    const insertIndex = Math.min(Math.max(index, 0), this.length);
    this.splice(insertIndex, 0, patternLayer);

    return patternLayer;
  }

  removeLayers(startIdx: number, endIdx: number): void {
    for (let i = endIdx; i >= startIdx; i--) {
      this.splice(i, 1);
    }
  }

  pushUpLayers(startIdx: number, endIdx: number): void {
    if (startIdx <= 0) return;
    const item = this.splice(startIdx - 1, 1)[0];
    this.splice(endIdx, 0, item);
  }

  pushDownLayers(startIdx: number, endIdx: number): void {
    if (endIdx >= this.length - 1) return;
    const item = this.splice(endIdx + 1, 1)[0];
    this.splice(startIdx, 0, item);
  }

  onLoadComplete(_context: TimeContext): void {
    // No-op for pattern layers
  }

  deepCopy(): PatternsLayerGroup {
    return new PatternsLayerGroup(this);
  }

  /** Get the maximum pattern index across all layers. */
  getMaxPattern(): number {
    let max = 0;
    for (const layer of this) {
      const ms = layer.getPatternData().getMaxSelected();
      if (ms > max) max = ms;
    }
    return max;
  }
}
